### PR TITLE
[search] Added villages.

### PR DESCRIPTION
**Pull request by mpimenov**

### CODE REVIEW

**Comment by mpimenov** [comment]
@vng @ygorshenin PTAL

File: indexer/ftypes_matcher.hpp [hunk 1]
```diff
@@ -84,6 +84,20 @@ class IsStreetChecker : public BaseChecker
   static IsStreetChecker const & Instance();
 };
 
+class IsVillageChecker : public BaseChecker
+{
+public:
+  IsVillageChecker();
```
Comment (ygorshenin) [inline indexer/ftypes_matcher.hpp]: Make this ctor private please.

Comment (mpimenov) [inline indexer/ftypes_matcher.hpp]: done

Comment (vng) [inline indexer/ftypes_matcher.hpp]: Конструктор можно в private секцию.

File: search/search_quality_tests/search_quality_tests.cpp [hunk 2]
```diff
@@ -219,7 +219,8 @@ int main(int argc, char * argv[])
   vector<double> responseTimes(queries.size());
   for (size_t i = 0; i < queries.size(); ++i)
   {
-    string const & query = queries[i];
+    // todo(@m) Add a bool flag to search with prefixes?
+    string const & query = queries[i] + " ";
```
Comment (ygorshenin) [inline search/search_quality_tests/search_quality_tests.cpp]: Current algorithm should work without space at the end of the query, right?

File: search/v2/geocoder.cpp [hunk 3]
```diff
@@ -609,6 +659,44 @@ void Geocoder::FillLocalitiesTable(MwmContext const & context)
...
+  size_t numVillages = 0;
+
+  for (auto & l : preLocalities)
+  {
+    FeatureType ft;
+    m_context->m_vector.GetByIndex(l.m_featureId, ft);
+
+    switch (m_model.GetSearchType(ft))
+    {
+    case SearchModel::SEARCH_TYPE_CITY:
```
Comment (ygorshenin) [inline search/v2/geocoder.cpp]: There should be a separate type - SEARCH_TYPE_VILLAGE, right?

File: search/v2/geocoder.cpp [hunk 4]
```diff
@@ -609,6 +659,44 @@ void Geocoder::FillLocalitiesTable(MwmContext const & context)
...
+    {
+      if (numVillages < kMaxNumVillages && ft.GetFeatureType() == feature::GEOM_POINT)
+      {
+        ++numVillages;
+        City village = l;
```
Comment (ygorshenin) [inline search/v2/geocoder.cpp]: It's better to create a type alias or a separate type.

File: search/v2/geocoder.cpp [hunk 5]
```diff
@@ -983,6 +1071,33 @@ void Geocoder::FindPaths()
...
+unique_ptr<coding::CompressedBitVector> Geocoder::LoadCategories(
+    MwmContext & context, vector<strings::UniString> const & categories)
+{
+  ASSERT(context.m_handle.IsAlive() && HasSearchIndex(context.m_value), ());
+
+  m_retrievalParams.m_tokens.resize(1);
+  m_retrievalParams.m_tokens[0].resize(1);
+  m_retrievalParams.m_prefixTokens.clear();
+
+  vector<unique_ptr<coding::CompressedBitVector>> list;
```
Comment (ygorshenin) [inline search/v2/geocoder.cpp]: vector list? Maybe it's worth to rename list to cbvs or to features?

File: search/v2/geocoder.cpp [hunk 6]
```diff
@@ -609,6 +659,44 @@ void Geocoder::FillLocalitiesTable(MwmContext const & context)
...
+        village.m_rect = MercatorBounds::RectByCenterXYAndSizeInMeters(
+            ft.GetCenter(), ftypes::GetRadiusByPopulation(ft.GetPopulation()));
+
+#if defined(DEBUG)
+        string name;
+        ft.GetName(StringUtf8Multilang::DEFAULT_CODE, name);
+        LOG(LDEBUG, ("Village =", name));
+#endif
+
+        m_cities[{l.m_startToken, l.m_endToken}].push_back(village);
```
Comment (ygorshenin) [inline search/v2/geocoder.cpp]: m_cities are filled with cities only once (from World.mwm), but villages are loaded from each mwm separately. Probably we need to remove villages from m_cities when we're moving from one mwm to another?

Comment (ygorshenin) [inline search/v2/geocoder.cpp]: One solution is to add a boolean flag (or tag) to struct City and then remove from m_cities only villages during iteration over mwms.

Comment (ygorshenin) [inline search/v2/geocoder.cpp]: Another is to implement a persistent tree.

File: search/search_quality_tests/search_quality_tests.cpp (hunk 2 above)
Comment (mpimenov) [inline search/search_quality_tests/search_quality_tests.cpp]: It works but won't the last token be treated as a prefix then? I don't want it to.

File: search/v2/geocoder.cpp (hunk 3 above)
Comment (mpimenov) [inline search/v2/geocoder.cpp]: I can add a new type if you want. I don't see much difference.

Comment (ygorshenin) [inline search/v2/geocoder.cpp]: It's possible to add cities from mwms then.

Comment (ygorshenin) [inline search/v2/geocoder.cpp]: I mean, you'll get not only villages but cities too.

Comment (mpimenov) [inline search/v2/geocoder.cpp]: There are no cities in m_villages, hence there are no cities in preLocalities.
Still, I'll think about it.

File: search/v2/geocoder.cpp (hunk 5 above)
Comment (mpimenov) [inline search/v2/geocoder.cpp]: done

File: search/v2/geocoder.cpp (hunk 6 above)
Comment (mpimenov) [inline search/v2/geocoder.cpp]: done

File: search/v2/geocoder.cpp (hunk 3 above)
Comment (mpimenov) [inline search/v2/geocoder.cpp]: done

File: search/v2/geocoder.cpp (hunk 4 above)
Comment (mpimenov) [inline search/v2/geocoder.cpp]: It's the only place where it is used.
1. Type allias would be called TVillage.
2. A separate type would infer code duplication.

**Comment by mpimenov** [comment]
PTAL

File: search/v2/geocoder.hpp [hunk 7]
```diff
@@ -239,9 +252,15 @@ class Geocoder : public my::Cancellable
   // Cache of street ids in mwms.
   map<MwmSet::MwmId, unique_ptr<coding::CompressedBitVector>> m_streetsCache;
 
-  // Streets in a currenly processed mwm.
+  // Cache of village ids in mwms.
+  map<MwmSet::MwmId, unique_ptr<coding::CompressedBitVector>> m_villagesCache;
```
Comment (vng) [inline search/v2/geocoder.hpp]: Вставить clear в Geocoder::ClearCaches().

**Comment by vng** [comment]
Учитывая ситуацию - LGTM, хотя как и со штатами/странами/городами, мне кажется что код можно упростить на 2 сущности:
- искать в определенном наборе mwm
- искать в районе пересечения CBV или ректа

**Comment by mpimenov** [comment]
Что значит "в определенном наборе mwm"?

Подождите мержить, ещё тесты.

**Comment by mpimenov** [comment]
@ygorshenin PTAL

**Comment by ygorshenin** [comment]
This CL has merge conflicts with master, please, rebase.

File: search/search_integration_tests/search_query_v2_test.cpp [hunk 8]
```diff
@@ -140,6 +152,7 @@ UNIT_TEST(SearchQueryV2_Smoke)
     builder.Add(*wonderlandCountry);
     builder.Add(*losAlamosCity);
     builder.Add(*mskCity);
+    builder.Add(*longPondVillage);
```
Comment (ygorshenin) [inline search/search_integration_tests/search_query_v2_test.cpp]: It's wrong to add village to the world file. Move it to the country file.

File: search/search_integration_tests/search_query_v2_test.cpp [hunk 9]
```diff
@@ -76,16 +76,12 @@ UNIT_TEST(SearchQueryV2_Smoke)
...
       make_shared<TestCity>(m2::PointD(10, 10), "Los Alamos", "en", 100 /* rank */);
   auto const mskCity = make_shared<TestCity>(m2::PointD(0, 0), "Moscow", "en", 100 /* rank */);
-  auto const busStop = make_shared<TestPOI>(m2::PointD(0, 0), "Bus stop", "en");
-  auto const tramStop = make_shared<TestPOI>(m2::PointD(0.0001, 0.0001), "Tram stop", "en");
-  auto const quantumTeleport1 =
-      make_shared<TestPOI>(m2::PointD(0.0002, 0.0002), "Quantum teleport 1", "en");
-  auto const quantumTeleport2 =
-      make_shared<TestPOI>(m2::PointD(10, 10), "Quantum teleport 2", "en");
-  auto const quantumCafe = make_shared<TestPOI>(m2::PointD(-0.0002, -0.0002), "Quantum cafe", "en");
+  auto const longPondVillage = make_shared<TestCity>(m2::PointD(15, 15), "Long Pond Village", "en", 10 /* rank */);
```
Comment (ygorshenin) [inline search/search_integration_tests/search_query_v2_test.cpp]: Please, implement a TestVillage class.

Comment (ygorshenin) [inline search/search_integration_tests/search_query_v2_test.cpp]: Ah, I see - you created it. Why not to use it here?

File: search/v2/geocoder.cpp [hunk 10]
```diff
@@ -689,6 +750,44 @@ void Geocoder::FillLocalitiesTable(MwmContext const & context)
...
+    switch (m_model.GetSearchType(ft))
+    {
+    case SearchModel::SEARCH_TYPE_VILLAGE:
+    {
+      if (numVillages < kMaxNumVillages && ft.GetFeatureType() == feature::GEOM_POINT)
+      {
+        ++numVillages;
+        City village = l;
+        village.m_rect = MercatorBounds::RectByCenterXYAndSizeInMeters(
+            ft.GetCenter(), ftypes::GetRadiusByPopulation(ft.GetPopulation()));
```
Comment (ygorshenin) [inline search/v2/geocoder.cpp]: Do not use GetCenter, use feature::GetCenter() instead.

File: search/v2/geocoder.cpp [hunk 11]
```diff
@@ -1092,6 +1191,33 @@ void Geocoder::FindPaths()
   });
 }
 
+unique_ptr<coding::CompressedBitVector> Geocoder::LoadCategories(
+    MwmContext & context, vector<strings::UniString> const & categories)
+{
+  ASSERT(context.m_handle.IsAlive() && HasSearchIndex(context.m_value), ());
```
Comment (ygorshenin) [inline search/v2/geocoder.cpp]: Please, separate this complex ASSERT() to two assertions. In the case of ASSERTion failure it's unclear what part of conjuction is false.

File: search/v2/geocoder.cpp (hunk 10 above)
Comment (ygorshenin) [inline search/v2/geocoder.cpp]: OK, for consistency with cities leave it as is, but note that ft.GetCenter() works only for points and may work incorrectly with complex shapes.

File: search/search_integration_tests/search_query_v2_test.cpp (hunk 9 above)
Comment (mpimenov) [inline search/search_integration_tests/search_query_v2_test.cpp]: I forgot to change it back, thanks.
done

File: search/search_integration_tests/search_query_v2_test.cpp (hunk 8 above)
Comment (mpimenov) [inline search/search_integration_tests/search_query_v2_test.cpp]: done

File: search/v2/geocoder.cpp (hunk 10 above)
Comment (mpimenov) [inline search/v2/geocoder.cpp]: done

File: search/v2/geocoder.cpp (hunk 11 above)
Comment (mpimenov) [inline search/v2/geocoder.cpp]: done

File: search/v2/geocoder.cpp (hunk 10 above)
Comment (mpimenov) [inline search/v2/geocoder.cpp]: Why? I've changed it to feature::GetCenter for cities too.

**Comment by mpimenov** [comment]
PTAL

File: search/v2/geocoder.cpp [hunk 12]
```diff
@@ -696,6 +754,44 @@ void Geocoder::FillLocalitiesTable(MwmContext const & context)
...
+  FillLocalityCandidates(m_villages.get(), kMaxNumVillages, preLocalities);
+
+  size_t numVillages = 0;
+
+  for (auto & l : preLocalities)
+  {
+    FeatureType ft;
+    m_context->m_vector.GetByIndex(l.m_featureId, ft);
+
+    switch (m_model.GetSearchType(ft))
```
Comment (ygorshenin) [inline search/v2/geocoder.cpp]: Do we need single statement switch? Why not to:
```c++
if (m_model.GetSearchType(ft) != SEARCH_TYPE_VILLAGE)
  continue;
...
```

Comment (mpimenov) [inline search/v2/geocoder.cpp]: doine

**Comment by mpimenov** [comment]
PTAL

**Comment by ygorshenin** [comment]
LGTM

**Pull request merged**